File: Scala/src/main/scala/TextbookTrack/Chapter04/BA4B.scala
package TextbookTrack.Chapter04

object BA4B {
  import TextbookTrack.Utils.readGeneticCode

  private val NucleotideComplements: Map[Char, Char] = Map('A' -> 'T', 'C' -> 'G', 'T' -> 'A', 'G' -> 'C')

  private def calcReverseComplement(dna: String): String = dna.reverseIterator.map(NucleotideComplements).mkString

  private def transcriptDna(dna: String): String = dna.map(nucleotide => if (nucleotide == 'T') 'U' else nucleotide)

  def findPeptideEncodingSubstrings(dna: String, peptide: String, geneticCode: Map[String, String]): List[String] = {
    def doesKMerTranslateToPeptide(pattern: String): Boolean =
      pattern.grouped(3).map(geneticCode.get).zip(peptide).forall{
        case (Some(aminoAcid), char) => aminoAcid == char.toString
        case (None, _) => false
      }

    val length: Int = dna.length
    val forwardRna: String = transcriptDna(dna)
    val reverseRna: String = transcriptDna(calcReverseComplement(dna))
    val k: Int = 3 * peptide.length
    forwardRna.sliding(k).zipWithIndex.foldLeft(List.empty[String]){
      case (encodingSubstrings, (pattern, ix)) =>
        val reversePattern: String = reverseRna.slice(length - ix - k, length - ix)
        if (doesKMerTranslateToPeptide(pattern) || doesKMerTranslateToPeptide(reversePattern))
          dna.slice(ix, ix + k) :: encodingSubstrings
        else encodingSubstrings
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val dna: String = reader.next()
    val peptide: String = reader.next()
    val geneticCode: Map[String, String] = readGeneticCode()
    val result: List[String] = findPeptideEncodingSubstrings(dna, peptide, geneticCode)
    result.foreach(println)
  }
}

File: Python/textbook_track/chapter04/ba4b.py
# Find Substrings of a Genome Encoding a Given Amino Acid String
import sys

from textbook_track.resources.utils import read_genetic_code

NUCLEOTIDE_COMPLEMENTS = {'A': 'T', 'C': 'G', 'T': 'A', 'G': 'C'}


def calc_reverse_complement(genome):
    return ''.join(map(NUCLEOTIDE_COMPLEMENTS.get, reversed(genome)))


def generate_codons(rna):
    for ix in range(len(rna) // 3):
        yield rna[3*ix:3*(ix+1)]


def transcript_dna(dna):
    return ''.join(('U' if nucleotide == 'T' else nucleotide for nucleotide in dna))


def does_k_mer_translate_to_peptide(k_mer, peptide, genetic_code):
    codons = generate_codons(k_mer)
    return all(genetic_code.get(codon) == amino_acid for codon, amino_acid in zip(codons, peptide))


def find_peptide_encoding_substrings(dna, peptide, genetic_code):
    length = len(dna)
    forward_rna = transcript_dna(dna)
    reverse_rna = transcript_dna(calc_reverse_complement(dna))
    k = 3 * len(peptide)
    encoding_substrings = []
    for ix in range(length-k+1):
        pattern = forward_rna[ix:ix+k]
        reverse_pattern = reverse_rna[length-ix-k:length-ix]
        if (does_k_mer_translate_to_peptide(pattern, peptide, genetic_code) or
            does_k_mer_translate_to_peptide(reverse_pattern, peptide, genetic_code)):
            encoding_substrings.append(dna[ix:ix+k])
    return encoding_substrings


def main():
    data = sys.stdin.read().splitlines()
    dna = data[0]
    peptide = data[1]
    genetic_code = read_genetic_code()
    result = find_peptide_encoding_substrings(dna, peptide, genetic_code)
    for substring in result:
        print(substring)


if __name__ == '__main__':
    main()
